lib/phase/permutation.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+const Hoek = require('hoek');
+const TinyTim = require('tinytim');
+const Hashr = require('hashr');
+const clone = require('clone');
+
+/**
+ * Generate all permutations of a given environment matrix
+ * @method generateMatrix
+ * @param  {Object}       options         Environment variable => Possible Values
+ * @param  {Object}       currentSettings Current selected fields
+ * @return {Array}                        List of permutations of the environment
+ */
+function generateMatrix(options, currentSettings) {
+    let permutations = [];
+
+    if (Object.keys(options).every((optionKey) => currentSettings[optionKey])) {
+        return [currentSettings];
+    }
+
+    Object.keys(options).forEach((optionKey) => {
+        if (currentSettings[optionKey]) {
+            return;
+        }
+        options[optionKey].forEach((option) => {
+            const newSettings = clone(currentSettings);
+
+            newSettings[optionKey] = option;
+            permutations = permutations.concat(generateMatrix(options, newSettings));
+        });
+    });
+
+    return permutations;
+}
+
+/**
+ * Permutation Phase
+ *  - Generate permutations for each job
+ * @method
+ * @param  {Object}   validatedDoc Document that went through functional validation
+ * @param  {Function} callback     Function to call when done (err, doc)
+ */
+module.exports = (validatedDoc, callback) => {
+    const doc = validatedDoc;
+
+    // Generate permutations for each job
+    Object.keys(doc.jobs).forEach((jobName) => {
+        const job = {
+            image: Hoek.reach(doc, `jobs.${jobName}.image`),
+            steps: Hoek.reach(doc, `jobs.${jobName}.steps`),
+            environment: Hoek.reach(doc, `jobs.${jobName}.environment`, {
+                default: {}
+            })
+        };
+        const matrix = Hoek.reach(doc, `jobs.${jobName}.matrix`, {
+            default: {}
+        });
+        const tasks = {};
+
+        generateMatrix(matrix, {}).forEach((permutation) => {
```

Comment:
LGTM Should we publish that as a generic module?  I'm still surprised there isn't one (or I can't seem to find it)